client/src/main.rs

```diff
@@ -20,52 +18,56 @@ fn main() {
     let metrics = crate::metrics::init();
 
     let dirs = directories::ProjectDirs::from("", "", "hypermine").unwrap();
-    let mut config = Config::load(&dirs);
-
-    if config.server.is_none() {
-        // spawn an in-process server
-        let socket =
-            UdpSocket::bind("[::1]:0".parse::<SocketAddr>().unwrap()).expect("binding socket");
-        config.server = Some(socket.local_addr().unwrap());
-
-        let certified_key = rcgen::generate_simple_self_signed(vec!["localhost".into()]).unwrap();
-        let key = certified_key.key_pair.serialize_der();
-        let cert = certified_key.cert.der().to_vec();
-        let sim_cfg = config.local_simulation.clone();
-
-        let save = dirs.data_local_dir().join(&config.save);
-        info!("using save file {}", save.display());
-        std::fs::create_dir_all(save.parent().unwrap()).unwrap();
-        let save = match Save::open(&save, config.local_simulation.chunk_size) {
-            Ok(x) => x,
-            Err(e) => {
-                error!("couldn't open save: {}", e);
-                return;
-            }
-        };
-
-        std::thread::spawn(move || {
-            let span = error_span!("server");
-            let _guard = span.enter();
-            if let Err(e) = server::run(
-                server::NetParams {
-                    certificate_chain: vec![CertificateDer::from(cert)],
-                    private_key: PrivatePkcs8KeyDer::from(key).into(),
-                    socket,
-                },
-                sim_cfg,
-                save,
-            ) {
-                eprintln!("{e:#}");
-                std::process::exit(1);
-            }
-        });
-    }
+    let config = Arc::new(Config::load(&dirs));
+
+    let net = match config.server {
+        None => {
+            // spawn an in-process server
+            let sim_cfg = config.local_simulation.clone();
+
+            let save = dirs.data_local_dir().join(&config.save);
+            info!("using save file {}", save.display());
+            std::fs::create_dir_all(save.parent().unwrap()).unwrap();
+            let save = match Save::open(&save, config.local_simulation.chunk_size) {
+                Ok(x) => x,
+                Err(e) => {
+                    error!("couldn't open save: {}", e);
+                    return;
+                }
+            };
+
+            let runtime = tokio::runtime::Builder::new_current_thread()
```

Comment:
I think part of why this doesn't work yet is that `runtime` is dropped and the end of this block, which is when `net` gets set to `handle`.

I believe the simplest fix is to spawn a thread for the runtime to live in. If the server code is to share a thread with the rest of the app, we'd need to see how hard it would be to share the thread with the blocking winit `run_app` function

Reply:
Yep, that was it. Thakns!